# Machine Learning With Spark ML
In this lab assignment, you will complete a project by going through the following steps:
1. Get the data.
2. Discover the data to gain insights.
3. Prepare the data for Machine Learning algorithms.
4. Select a model and train it.
5. Fine-tune your model.
6. Present your solution.

As a dataset, we use the California Housing Prices dataset from the StatLib repository. This dataset was based on data from the 1990 California census. The dataset has the following columns
1. `longitude`: a measure of how far west a house is (a higher value is farther west)
2. `latitude`: a measure of how far north a house is (a higher value is farther north)
3. `housing_,median_age`: median age of a house within a block (a lower number is a newer building)
4. `total_rooms`: total number of rooms within a block
5. `total_bedrooms`: total number of bedrooms within a block
6. `population`: total number of people residing within a block
7. `households`: total number of households, a group of people residing within a home unit, for a block
8. `median_income`: median income for households within a block of houses
9. `median_house_value`: median house value for households within a block
10. `ocean_proximity`: location of the house w.r.t ocean/sea

---
# 1. Get the data
Let's start the lab by loading the dataset. The can find the dataset at `data/housing.csv`. To infer column types automatically, when you are reading the file, you need to set `inferSchema` to true. Moreover enable the `header` option to read the columns' name from the file.

In [1]:
// TODO: Replace <FILL IN> with appropriate code

val housing = spark.read.format("csv").option("header","true").option("inferSchema","true").load("data/housing.csv")

housing = [longitude: double, latitude: double ... 8 more fields]


[longitude: double, latitude: double ... 8 more fields]

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

---
# 2. Discover the data to gain insights
Now it is time to take a look at the data. In this step we are going to take a look at the data a few different ways:
* See the schema and dimension of the dataset
* Look at the data itself
* Statistical summary of the attributes
* Breakdown of the data by the categorical attribute variable
* Find the correlation among different attributes
* Make new attributes by combining existing attributes

## 2.1. Schema and dimension
Print the schema of the dataset

In [2]:
// TODO: Replace <FILL IN> with appropriate code

println(housing.schema)

StructType(StructField(longitude,DoubleType,true), StructField(latitude,DoubleType,true), StructField(housing_median_age,DoubleType,true), StructField(total_rooms,DoubleType,true), StructField(total_bedrooms,DoubleType,true), StructField(population,DoubleType,true), StructField(households,DoubleType,true), StructField(median_income,DoubleType,true), StructField(median_house_value,DoubleType,true), StructField(ocean_proximity,StringType,true))


Print the number of records in the dataset.

In [3]:
// TODO: Replace <FILL IN> with appropriate code

println("Number of rows are: " + housing.count)

Number of rows are: 20640


## 2.2. Look at the data
Print the first five records of the dataset.

In [4]:
// TODO: Replace <FILL IN> with appropriate code

housing.take(5).foreach(println)

[-122.23,37.88,41.0,880.0,129.0,322.0,126.0,8.3252,452600.0,NEAR BAY]
[-122.22,37.86,21.0,7099.0,1106.0,2401.0,1138.0,8.3014,358500.0,NEAR BAY]
[-122.24,37.85,52.0,1467.0,190.0,496.0,177.0,7.2574,352100.0,NEAR BAY]
[-122.25,37.85,52.0,1274.0,235.0,558.0,219.0,5.6431,341300.0,NEAR BAY]
[-122.25,37.85,52.0,1627.0,280.0,565.0,259.0,3.8462,342200.0,NEAR BAY]


Print the number of records with population more than 10000.

In [5]:
// TODO: Replace <FILL IN> with appropriate code

println("Number of records with population bigger than 10000 are: " + housing.where("population > 10000").count)

Number of records with population bigger than 10000 are: 23


## 2.3. Statistical summary
Print a summary of the table statistics for the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. You can use the `describe` command.

In [6]:
// TODO: Replace <FILL IN> with appropriate code

//hivecontext.sql("describe formatted table_name").collect().foreach(println)


housing.describe("housing_median_age","total_rooms","median_house_value","population").show()

+-------+------------------+------------------+------------------+------------------+
|summary|housing_median_age|       total_rooms|median_house_value|        population|
+-------+------------------+------------------+------------------+------------------+
|  count|             20640|             20640|             20640|             20640|
|   mean|28.639486434108527|2635.7630813953488|206855.81690891474|1425.4767441860465|
| stddev| 12.58555761211163|2181.6152515827944|115395.61587441359|  1132.46212176534|
|    min|               1.0|               2.0|           14999.0|               3.0|
|    max|              52.0|           39320.0|          500001.0|           35682.0|
+-------+------------------+------------------+------------------+------------------+



Print the maximum age (`housing_median_age`), the minimum number of rooms (`total_rooms`), and the average of house values (`median_house_value`).

In [7]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.sql.functions._

println("housing_median_age, total_rooms, median_house_value")
housing.agg(max("housing_median_age"),min("total_rooms"), avg("median_house_value")).collect

housing_median_age, total_rooms, median_house_value


Array([52.0,2.0,206855.81690891474])

## 2.4. Breakdown the data by categorical data
Print the number of houses in different areas (`ocean_proximity`), and sort them in descending order.

In [8]:
// TODO: Replace <FILL IN> with appropriate code

housing.groupBy("ocean_proximity").agg(count("ocean_proximity").alias("groupAgg")).orderBy(desc("groupAgg")).show()

+---------------+--------+
|ocean_proximity|groupAgg|
+---------------+--------+
|      <1H OCEAN|    9136|
|         INLAND|    6551|
|     NEAR OCEAN|    2658|
|       NEAR BAY|    2290|
|         ISLAND|       5|
+---------------+--------+



Print the average value of the houses (`median_house_value`) in different areas (`ocean_proximity`), and call the new column `avg_value` when print it.

In [9]:
// TODO: Replace <FILL IN> with appropriate code

housing.groupBy("ocean_proximity").agg(avg("median_house_value").alias("avg_value")).orderBy(desc("avg_value")).show()

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|       NEAR BAY|259212.31179039303|
|     NEAR OCEAN|249433.97742663656|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



Rewrite the above question in SQL.

In [10]:
// TODO: Replace <FILL IN> with appropriate code

housing.createOrReplaceTempView("df")
spark.sql("select ocean_proximity, avg(median_house_value) as avg_value from df group by ocean_proximity").show()

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



## 2.5. Correlation among attributes
Print the correlation among the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. To do so, first you need to put these attributes into one vector. Then, compute the standard correlation coefficient (Pearson) between every pair of attributes in this new vector. To make a vector of these attributes, you can use the `VectorAssembler` Transformer.

In [11]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.VectorAssembler

case class Nums(val1: Long,val2: Long,val3: Long,val4: Long)

// Maybe not needed
val numsDF = housing.select("housing_median_age","total_rooms","median_house_value", "population")
val va = new VectorAssembler().setInputCols(Array("housing_median_age","total_rooms","median_house_value", "population")).setOutputCol("CombVector")

val housingAttrs = va.transform(numsDF)

housingAttrs.show(5)

+------------------+-----------+------------------+----------+--------------------+
|housing_median_age|total_rooms|median_house_value|population|          CombVector|
+------------------+-----------+------------------+----------+--------------------+
|              41.0|      880.0|          452600.0|     322.0|[41.0,880.0,45260...|
|              21.0|     7099.0|          358500.0|    2401.0|[21.0,7099.0,3585...|
|              52.0|     1467.0|          352100.0|     496.0|[52.0,1467.0,3521...|
|              52.0|     1274.0|          341300.0|     558.0|[52.0,1274.0,3413...|
|              52.0|     1627.0|          342200.0|     565.0|[52.0,1627.0,3422...|
+------------------+-----------+------------------+----------+--------------------+
only showing top 5 rows



defined class Nums
numsDF = [housing_median_age: double, total_rooms: double ... 2 more fields]
va = vecAssembler_00aa683ef334
housingAttrs = [housing_median_age: double, total_rooms: double ... 3 more fields]


[housing_median_age: double, total_rooms: double ... 3 more fields]

In [12]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.linalg.Matrix
import org.apache.spark.ml.stat.Correlation
import org.apache.spark.sql.Row

val Row(coeff: Matrix) = Correlation.corr(housingAttrs,"CombVector").head

println(s"The standard correlation coefficient:\n ${coeff}")

The standard correlation coefficient:
 1.0                   -0.36126220122231784  0.10562341249318154   -0.2962442397735293   
-0.36126220122231784  1.0                   0.13415311380654338   0.8571259728659772    
0.10562341249318154   0.13415311380654338   1.0                   -0.02464967888891235  
-0.2962442397735293   0.8571259728659772    -0.02464967888891235  1.0                   


coeff = 


1.0                   -0.36126220122231784  0.10562341249318154   -0.2962442397735293
-0.36126220122231784  1.0                   0.13415311380654338   0.8571259728659772
0.10562341249318154   0.13415311380654338   1.0                   -0.02464967888891235
-0.2962442397735293   0.8571259728659772    -0.02464967888891235  1.0


## 2.6. Combine and make new attributes
Now, let's try out various attribute combinations. In the given dataset, the total number of rooms in a block is not very useful, if we don't know how many households there are. What we really want is the number of rooms per household. Similarly, the total number of bedrooms by itself is not very useful, and we want to compare it to the number of rooms. And the population per household seems like also an interesting attribute combination to look at. To do so, add the three new columns to the dataset as below. We will call the new dataset the `housingExtra`.
```
rooms_per_household = total_rooms / households
bedrooms_per_room = total_bedrooms / total_rooms
population_per_household = population / households
```

In [13]:
// TODO: Replace <FILL IN> with appropriate code
// Not sure about this one, he wants us to do it differently

val housingCol1 = housing.withColumn("rooms_per_household", housing("total_rooms") / housing("households"))
val housingCol2 = housingCol1.withColumn("bedrooms_per_room", housing("total_bedrooms") / housing("total_rooms"))
val housingColExtra = housingCol2.withColumn("population_per_household", housing("population") / housing("households"))


housingColExtra.select("rooms_per_household", "bedrooms_per_room", "population_per_household").show(5)

+-------------------+-------------------+------------------------+
|rooms_per_household|  bedrooms_per_room|population_per_household|
+-------------------+-------------------+------------------------+
|  6.984126984126984|0.14659090909090908|      2.5555555555555554|
|  6.238137082601054|0.15579659106916466|       2.109841827768014|
|  8.288135593220339|0.12951601908657123|      2.8022598870056497|
| 5.8173515981735155|0.18445839874411302|       2.547945205479452|
|  6.281853281853282| 0.1720958819913952|      2.1814671814671813|
+-------------------+-------------------+------------------------+
only showing top 5 rows



housingCol1 = [longitude: double, latitude: double ... 9 more fields]
housingCol2 = [longitude: double, latitude: double ... 10 more fields]
housingColExtra = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

---
## 3. Prepare the data for Machine Learning algorithms
Before going through the Machine Learning steps, let's first rename the label column from `median_house_value` to `label`.

In [14]:
// TODO: Replace <FILL IN> with appropriate code

val renamedHousing = housingColExtra.withColumnRenamed("median_house_value","label")
renamedHousing.take(5)

renamedHousing = [longitude: double, latitude: double ... 11 more fields]


Array([-122.23,37.88,41.0,880.0,129.0,322.0,126.0,8.3252,452600.0,NEAR BAY,6.984126984126984,0.14659090909090908,2.5555555555555554], [-122.22,37.86,21.0,7099.0,1106.0,2401.0,1138.0,8.3014,358500.0,NEAR BAY,6.238137082601054,0.15579659106916466,2.109841827768014], [-122.24,37.85,52.0,1467.0,190.0,496.0,177.0,7.2574,352100.0,NEAR BAY,8.288135593220339,0.12951601908657123,2.8022598870056497], [-122.25,37.85,52.0,1274.0,235.0,558.0,219.0,5.6431,341300.0,NEAR BAY,5.8173515981735155,0.18445839874411302,2.547945205479452], [-122.25,37.85,52.0,1627.0,280.0,565.0,259.0,3.8462,342200.0,NEAR BAY,6.281853281853282,0.1720958819913952,2.1814671814671813])

Now, we want to separate the numerical attributes from the categorical attribute (`ocean_proximity`) and keep their column names in two different lists. Moreover, sice we don't want to apply the same transformations to the predictors (features) and the label, we should remove the label attribute from the list of predictors. 

In [15]:
// label columns
val colLabel = "label"

// categorical columns
val colCat = "ocean_proximity"

// numerical columns
val colNum = renamedHousing.columns.filter(_ != colLabel).filter(_ != colCat)

colLabel = label
colCat = ocean_proximity
colNum = Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)


Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)

## 3.1. Prepare continuse attributes
### Data cleaning
Most Machine Learning algorithms cannot work with missing features, so we should take care of them. As a first step, let's find the columns with missing values in the numerical attributes. To do so, we can print the number of missing values of each continues attributes, listed in `colNum`.

In [16]:
// TODO: Replace <FILL IN> with appropriate code

for (c <- colNum) {
    println("Column " + c + " has " + renamedHousing.filter(c + " is null").count + " missing values.")
}

Column longitude has 0 missing values.
Column latitude has 0 missing values.
Column housing_median_age has 0 missing values.
Column total_rooms has 0 missing values.
Column total_bedrooms has 207 missing values.
Column population has 0 missing values.
Column households has 0 missing values.
Column median_income has 0 missing values.
Column rooms_per_household has 0 missing values.
Column bedrooms_per_room has 207 missing values.
Column population_per_household has 0 missing values.


As we observerd above, the `total_bedrooms` and `bedrooms_per_room` attributes have some missing values. One way to take care of missing values is to use the `Imputer` Transformer, which completes missing values in a dataset, either using the mean or the median of the columns in which the missing values are located. To use it, you need to create an `Imputer` instance, specifying that you want to replace each attribute's missing values with the "median" of that attribute.

In [17]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.Imputer

val imputer = new Imputer().setStrategy("median")
        .setInputCols(Array("total_bedrooms","bedrooms_per_room"))
        .setOutputCols(Array("total_bedrooms","bedrooms_per_room"))
val imputedHousing = imputer.fit(renamedHousing).transform(renamedHousing)

imputedHousing.select("total_bedrooms", "bedrooms_per_room").show(5)

+--------------+-------------------+
|total_bedrooms|  bedrooms_per_room|
+--------------+-------------------+
|         129.0|0.14659090909090908|
|        1106.0|0.15579659106916466|
|         190.0|0.12951601908657123|
|         235.0|0.18445839874411302|
|         280.0| 0.1720958819913952|
+--------------+-------------------+
only showing top 5 rows



imputer = imputer_343044857664
imputedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

### Scaling
One of the most important transformations you need to apply to your data is feature scaling. With few exceptions, Machine Learning algorithms don't perform well when the input numerical attributes have very different scales. This is the case for the housing data: the total number of rooms ranges from about 6 to 39,320, while the median incomes only range from 0 to 15. Note that scaling the label attribues is generally not required.

One way to get all attributes to have the same scale is to use standardization. In standardization, for each value, first it subtracts the mean value (so standardized values always have a zero mean), and then it divides by the variance so that the resulting distribution has unit variance. To do this, we can use the `StandardScaler` Estimator. To use `StandardScaler`, again we need to convert all the numerical attributes into a big vectore of features using `VectorAssembler`, and then call `StandardScaler` on that vactor.

In [18]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

val va = new VectorAssembler().setInputCols(colNum).setOutputCol("features1")
val featuredHousing = va.transform(imputedHousing)

val scaler = new StandardScaler().setInputCol("features1").setOutputCol("scales")
val scaledHousing = scaler.fit(featuredHousing).transform(featuredHousing)

scaledHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|           features1|              scales|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|[-122.23,37.88,41...|[-61.007269596069...|
|  -122.22|   37.86|              21.0|     7099.0|        1

va = vecAssembler_e50634a34f40
featuredHousing = [longitude: double, latitude: double ... 12 more fields]
scaler = stdScal_b1f2c235e937
scaledHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

## 3.2. Prepare categorical attributes
After imputing and scaling the continuse attributes, we should take care of the categorical attributes. Let's first print the number of distict values of the categirical attribute `ocean_proximity`.

In [19]:
// TODO: Replace <FILL IN> with appropriate code
import org.apache.spark.sql.functions.countDistinct

renamedHousing.select(countDistinct("ocean_proximity").alias("Distinct ocean proximity")).show

+------------------------+
|Distinct ocean proximity|
+------------------------+
|                       5|
+------------------------+



### String indexer
Most Machine Learning algorithms prefer to work with numbers. So let's convert the categorical attribute `ocean_proximity` to numbers. To do so, we can use the `StringIndexer` that encodes a string column of labels to a column of label indices. The indices are in [0, numLabels), ordered by label frequencies, so the most frequent label gets index 0.

In [20]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.StringIndexer

val indexer = new StringIndexer().setInputCol("ocean_proximity").setOutputCol("labelOP")
val idxHousing = indexer.fit(renamedHousing).transform(renamedHousing)

idxHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+-------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|labelOP|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+-------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|    3.0|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|358500.0|       NEAR BAY|  6.238137082601054|0.15579659106916466|       2.109841827768014|    

indexer = strIdx_0c4d19d56d51
idxHousing = [longitude: double, latitude: double ... 12 more fields]


[longitude: double, latitude: double ... 12 more fields]

Now we can use this numerical data in any Machine Learning algorithm. You can look at the mapping that this encoder has learned using the `labels` method: "<1H OCEAN" is mapped to 0, "INLAND" is mapped to 1, etc.

In [21]:
val ind = indexer.fit(renamedHousing).labels

ind = Array(<1H OCEAN, INLAND, NEAR OCEAN, NEAR BAY, ISLAND)


Array(<1H OCEAN, INLAND, NEAR OCEAN, NEAR BAY, ISLAND)

### One-hot encoding
Now, convert the label indices built in the last step into one-hot vectors. To do this, you can take advantage of the `OneHotEncoderEstimator` Estimator.

In [22]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.OneHotEncoderEstimator

val encoder = new OneHotEncoderEstimator().setInputCols(Array("labelOP")).setOutputCols(Array("OneHotlabelOP"))
val ohHousing = encoder.fit(idxHousing).transform(idxHousing)

//ohHousing.select(ohHousing("OneHotlabelOP")).distinct.show

ohHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+-------+-------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|labelOP|OneHotlabelOP|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+-------+-------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|    3.0|(4,[3],[1.0])|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|358500.0|       NEAR BAY|  6.238137082

encoder = oneHotEncoder_de0d7bad2eb5
ohHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

---
# 4. Pipeline
As you can see, there are many data transformation steps that need to be executed in the right order. For example, you called the `Imputer`, `VectorAssembler`, and `StandardScaler` from left to right. However, we can use the `Pipeline` class to define a sequence of Transformers/Estimators, and run them in order. A `Pipeline` is an `Estimator`, thus, after a Pipeline's `fit()` method runs, it produces a `PipelineModel`, which is a `Transformer`.

Now, let's create a pipeline called `numPipeline` to call the numerical transformers you built above (`imputer`, `va`, and `scaler`) in the right order from left to right, as well as a pipeline called `catPipeline` to call the categorical transformers (`indexer` and `encoder`). Then, put these two pipelines `numPipeline` and `catPipeline` into one pipeline.

In [23]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.{Pipeline, PipelineModel, PipelineStage}

val numPipeline = new Pipeline().setStages(Array(imputer, va, scaler))
val catPipeline = new Pipeline().setStages(Array(indexer, encoder))
val pipeline = new Pipeline().setStages(Array(numPipeline, catPipeline))
val newHousing = pipeline.fit(renamedHousing).transform(renamedHousing)

newHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+-------+-------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|           features1|              scales|labelOP|OneHotlabelOP|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+-------+-------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|[-122.23,37.88,41...|[-61.00726959606

numPipeline = pipeline_fe3ff4a8d96d
catPipeline = pipeline_43c9f1e08ffc
pipeline = pipeline_69a350aee45c
newHousing = [longitude: double, latitude: double ... 15 more fields]


[longitude: double, latitude: double ... 15 more fields]

Now, use `VectorAssembler` to put all attributes of the final dataset `newHousing` into a big vector, and call the new column `features`.

In [24]:
// TODO: Replace <FILL IN> with appropriate code


val va2 = new VectorAssembler().setInputCols(Array("scales", "OneHotlabelOP")).setOutputCol("features")
val dataset = va2.transform(newHousing).select("features", "label")

dataset.show(5)

+--------------------+--------+
|            features|   label|
+--------------------+--------+
|[-61.007269596069...|452600.0|
|[-61.002278409814...|358500.0|
|[-61.012260782324...|352100.0|
|[-61.017251968579...|341300.0|
|[-61.017251968579...|342200.0|
+--------------------+--------+
only showing top 5 rows



va2 = vecAssembler_4effaeee9497
dataset = [features: vector, label: double]


[features: vector, label: double]

---
# 5. Make a model
Here we going to make four different regression models:
* Linear regression model
* Decission tree regression
* Random forest regression
* Gradient-booster forest regression

But, before giving the data to train a Machine Learning model, let's first split the data into training dataset (`trainSet`) with 80% of the whole data, and test dataset (`testSet`) with 20% of it.

In [25]:
// TODO: Replace <FILL IN> with appropriate code

val Array(trainSet, testSet) = dataset.randomSplit(Array(0.8, 0.2), seed = 12345)

trainSet = [features: vector, label: double]
testSet = [features: vector, label: double]


[features: vector, label: double]

## 5.1. Linear regression model
Now, train a Linear Regression model using the `LinearRegression` class. Then, print the coefficients and intercept of the model, as well as the summary of the model over the training set by calling the `summary` method.

In [26]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.LinearRegression

// train the model
val lr = new LinearRegression()
      .setFeaturesCol("features")
      .setLabelCol("label")
      .setRegParam(0.1)
      .setMaxIter(10)
val lrModel = lr.fit(trainSet)
val trainingSummary = lrModel.summary

println(s"Coefficients: ${lrModel.coefficients}, Intercept: ${lrModel.intercept}")
println(s"RMSE: ${trainingSummary.rootMeanSquaredError}")

Coefficients: [-55095.6537644947,-55748.680215075125,13542.532220737881,7895.340713072312,189.47111749539883,-46107.67645467558,42891.12157250382,78168.01540611999,6380.730964628803,17047.903461882313,-29.978998453907852,-132534.62334231552,-166737.82849368345,-128202.09254421295,-136238.00415951438], Intercept: -2283214.512355055
RMSE: 67768.46304870817


lr = linReg_4e4dcdfe302d
lrModel = linReg_4e4dcdfe302d
trainingSummary = org.apache.spark.ml.regression.LinearRegressionTrainingSummary@6179bef


org.apache.spark.ml.regression.LinearRegressionTrainingSummary@6179bef

Now, use `RegressionEvaluator` to measure the root-mean-square-erroe (RMSE) of the model on the test dataset.

In [27]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.evaluation.RegressionEvaluator

// make predictions on the test data
val predictions = lrModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setMetricName("rmse").setPredictionCol("prediction").setLabelCol("label")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|180099.72220010776|111400.0|[-62.020480405854...|
|206764.39850573335| 50800.0|[-62.005506847089...|
|114413.63609048631| 73200.0|[-62.005506847089...|
|144777.80357633065| 70000.0|[-61.985542102068...|
| 94254.35597636597| 74600.0|[-61.985542102068...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 68407.6585983051


predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_a757fa820a99
rmse = 68407.6585983051


68407.6585983051

## 5.2. Decision tree regression
Repeat what you have done on Regression Model to build a Decision Tree model. Use the `DecisionTreeRegressor` to make a model and then measure its RMSE on the test dataset.

In [28]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.DecisionTreeRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val dt = new DecisionTreeRegressor().setLabelCol("label").setFeaturesCol("features")

// train the model
val dtModel = dt.fit(trainSet)

// make predictions on the test data
val predictions = dtModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setMetricName("rmse").setPredictionCol("prediction").setLabelCol("label")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|132095.55555555556|111400.0|[-62.020480405854...|
| 167794.4568788501| 50800.0|[-62.005506847089...|
| 167794.4568788501| 73200.0|[-62.005506847089...|
|132095.55555555556| 70000.0|[-61.985542102068...|
|137658.92252803262| 74600.0|[-61.985542102068...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 68684.58653764601


dt = dtr_7d34b14cbdf8
dtModel = DecisionTreeRegressionModel (uid=dtr_7d34b14cbdf8) of depth 5 with 63 nodes
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_0d8df1daf4f3
rmse = 68684.58653764601


68684.58653764601

## 5.3. Random forest regression
Let's try the test error on a Random Forest Model. Youcan use the `RandomForestRegressor` to make a Random Forest model.

In [29]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.RandomForestRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val rf = new RandomForestRegressor().setLabelCol("label").setFeaturesCol("features")

// train the model
val rfModel = rf.fit(trainSet)

// make predictions on the test data
val predictions = rfModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setMetricName("rmse").setPredictionCol("prediction").setLabelCol("label")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 172238.5170535912|111400.0|[-62.020480405854...|
|218572.07153806346| 50800.0|[-62.005506847089...|
| 184479.8494104938| 73200.0|[-62.005506847089...|
|153099.14507029785| 70000.0|[-61.985542102068...|
|155421.01444762942| 74600.0|[-61.985542102068...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 66269.53792130198


rf = rfr_a7df260de39e
rfModel = RandomForestRegressionModel (uid=rfr_a7df260de39e) with 20 trees
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_1e5cde16868b
rmse = 66269.53792130198


66269.53792130198

## 5.4. Gradient-boosted tree regression
Fianlly, we want to build a Gradient-boosted Tree Regression model and test the RMSE of the test data. Use the `GBTRegressor` to build the model.

In [30]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.GBTRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val gb = new GBTRegressor().setLabelCol("label")
      .setFeaturesCol("features")
      .setMaxIter(10)

// train the model
val gbModel = gb.fit(trainSet)

// make predictions on the test data
val predictions = gbModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setMetricName("rmse").setPredictionCol("prediction").setLabelCol("label")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 81035.09601629924|111400.0|[-62.020480405854...|
|106514.97364352363| 50800.0|[-62.005506847089...|
| 143134.9288207957| 73200.0|[-62.005506847089...|
| 92478.25391711286| 70000.0|[-61.985542102068...|
|103622.40833211431| 74600.0|[-61.985542102068...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 59934.29928558595


gb = gbtr_417e45551f7b
gbModel = GBTRegressionModel (uid=gbtr_417e45551f7b) with 10 trees
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_ca968b4e6276
rmse = 59934.29928558595


59934.29928558595

---
# 6. Hyperparameter tuning
An important task in Machie Learning is model selection, or using data to find the best model or parameters for a given task. This is also called tuning. Tuning may be done for individual Estimators such as LinearRegression, or for entire Pipelines which include multiple algorithms, featurization, and other steps. Users can tune an entire Pipeline at once, rather than tuning each element in the Pipeline separately. MLlib supports model selection tools, such as `CrossValidator`. These tools require the following items:
* Estimator: algorithm or Pipeline to tune (`setEstimator`)
* Set of ParamMaps: parameters to choose from, sometimes called a "parameter grid" to search over (`setEstimatorParamMaps`)
* Evaluator: metric to measure how well a fitted Model does on held-out test data (`setEvaluator`)

`CrossValidator` begins by splitting the dataset into a set of folds, which are used as separate training and test datasets. For example with `k=3` folds, `CrossValidator` will generate 3 (training, test) dataset pairs, each of which uses 2/3 of the data for training and 1/3 for testing. To evaluate a particular `ParamMap`, `CrossValidator` computes the average evaluation metric for the 3 Models produced by fitting the Estimator on the 3 different (training, test) dataset pairs. After identifying the best `ParamMap`, `CrossValidator` finally re-fits the Estimator using the best ParamMap and the entire dataset.

Below, use the `CrossValidator` to select the best Random Forest model. To do so, you need to define a grid of parameters. Let's say we want to do the search among the different number of trees (1, 5, and 10), and different tree depth (5, 10, and 15).

In [31]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.RegressionEvaluator
import org.apache.spark.ml.tuning.CrossValidator

val paramGrid = new ParamGridBuilder().addGrid(rf.numTrees, Array(1, 5, 10)).addGrid(rf.maxDepth, Array(5, 10, 15)).build()

val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val cv = new CrossValidator().setEstimator(rf).setEvaluator(evaluator).setEstimatorParamMaps(paramGrid).setNumFolds(3)
val cvModel = cv.fit(trainSet)

val predictions = cvModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+-----------------+--------+--------------------+
|       prediction|   label|            features|
+-----------------+--------+--------------------+
|         123470.0|111400.0|[-62.020480405854...|
|         142560.0| 50800.0|[-62.005506847089...|
|         103525.0| 73200.0|[-62.005506847089...|
| 74192.9869385001| 70000.0|[-61.985542102068...|
|98243.28205128205| 74600.0|[-61.985542102068...|
+-----------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 53627.903370110995


paramGrid = 


Array({
	rfr_a7df260de39e-maxDepth: 5,
	rfr_a7df260de39e-numTrees: 1
}, {
	rfr_a7df260de39e-maxDepth: 10,
	rfr_a7df260de39e-numTrees: 1
}, {
	rfr_a7df260de39e-maxDepth: 15,
	rfr_a7df260de39e-numTrees: 1
}, {
	rfr_a7df260de39e-maxDepth: 5,
	rfr_a7df260de39e-numTrees: 5
}, {
	rfr_a7df260de39e-maxDepth: 10,
	rfr_a7df260de39e-numTrees: 5
}, {
	rfr_a7df260de39e-maxDepth: 15,
	rfr_a7df260de39e-numTrees: 5
}, {
	rfr_a7df260de39e-maxDepth: 5,
	rfr_a7df260de39e-numTrees: 10
}, {
	rfr_a7df260de39e-maxDepth: 10,
	rfr_a7df260de39e-numTrees: 10
}, {
	rfr_a7df260de39e-maxDepth: 15,
	rfr_a7df...


---
# 7. Custom transformer
At the end of part two, we added extra columns to the `housing` dataset. Here, we are going to implement a Transformer to do the same task. The Transformer should take the name of two input columns `inputCol1` and `inputCol2`, as well as the name of ouput column `outputCol`. It, then, computes `inputCol1` divided by `inputCol2`, and adds its result as a new column to the dataset. The details of the implemeting a custom Tranfomer is explained [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types). Please read it before before starting to implement it.

First, define the given parameters of the Transformer and implement a method to validate their schemas (`StructType`).

In [32]:
import org.apache.spark.sql.types.{StructField, StructType, DoubleType}
import org.apache.spark.ml.param.{ParamMap, Param, Params}



trait MyParams extends Params {
    final val inputCol1 = new Param[String](this, "inputCol1", "The first input Column")
    final val inputCol2 = new Param[String](this, "inputCol2", "The second input Column")
    final val outputCol = new Param[String](this, "outputCol", "The output column")
    
  protected def validateAndTransformSchema(schema: StructType): StructType = {      
      val inputfield = schema.fieldIndex("inputCol1")
      val field = schema.fields(inputfield)
      
      val inputfield2 = schema.fieldIndex("inputCol2")
      val field2 = schema.fields(inputfield2)
      
      if (field.dataType != DoubleType || field.dataType != DoubleType)
      {
          throw new Exception ("Input type did not match input type DoubleType")
      }
      
      return schema.add(StructField("outputCol", DoubleType, false))
      
  }
}


defined trait MyParams


Then, extend the class `Transformer`, and implement its setter functions for the input and output columns, and call then `setInputCol1`, `setInputCol2`, and `setOutputCol`. Morever, you need to override the methods `copy`, `transformSchema`, and the `transform`. The details of what you need to cover in these methods is given [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types).

In [33]:
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.ml.Transformer
import org.apache.spark.ml.param.{ParamMap, Param, Params}
import org.apache.spark.sql.{DataFrame, Dataset}
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.functions.{col, udf}


val housingCol1 = housing.withColumn("rooms_per_household", housing("total_rooms") / housing("households"))
val housingCol2 = housingCol1.withColumn("bedrooms_per_room", housing("total_bedrooms") / housing("total_rooms"))
val housingColExtra = housingCol2.withColumn("population_per_household", housing("population") / housing("households"))





class MyTransformer(override val uid: String) extends Transformer with MyParams {
    def this() = this(Identifiable.randomUID("configurablewordcount"))
    
    def setInputCol1(value: String): this.type = set(inputCol1, value)
    
    def setInputCol2(value: String): this.type = set(inputCol2, value)
    
    def setOutputCol(value: String): this.type = set(outputCol, value)

    override def copy(extra: ParamMap): MyTransformer = {
    defaultCopy(extra)
    }

    override def transformSchema(schema: StructType) : StructType = {
        validateAndTransformSchema(schema)
    }
    
    override def transform(dataset: Dataset[_]): DataFrame = {
        
        val col1 = dataset.select((col($(inputCol1)) / col($(inputCol2))).as($(outputCol)))
        col1
  }
}

housingCol1 = [longitude: double, latitude: double ... 9 more fields]
housingCol2 = [longitude: double, latitude: double ... 10 more fields]
housingColExtra = [longitude: double, latitude: double ... 11 more fields]
defined class MyTransformer


[longitude: double, latitude: double ... 11 more fields]

Now, an instance of `MyTransformer`, and set the input columns `total_rooms` and `households`, and the output column `rooms_per_household` and run it over the `housing` dataset.

In [34]:
val myTransformer = new MyTransformer().setInputCol1("total_rooms").setInputCol2("households").setOutputCol("rooms_per_household")

val myDataset = myTransformer.transform(housing).select("rooms_per_household").show(5)

+-------------------+
|rooms_per_household|
+-------------------+
|  6.984126984126984|
|  6.238137082601054|
|  8.288135593220339|
| 5.8173515981735155|
|  6.281853281853282|
+-------------------+
only showing top 5 rows



myTransformer = configurablewordcount_1cbc1d014aa7


myDataset: Unit = ()


configurablewordcount_1cbc1d014aa7

---
# 8. Custom estimator (predictor)
Now, it's time to implement your own linear regression with gradient descent algorithm as a brand new Estimator. The whole code of the Estimator is given to you, and you do not need to implement anything. It is just a sample that shows how to build a custom Estimator.

The gradient descent update for linear regression is:
$$
w_{i+1} = w_{i} - \alpha_{i} \sum\limits_{j=1}^n (w_i^\top x_j - y_j)x_j
$$

where $i$ is the iteration number of the gradient descent algorithm, and $j$ identifies the observation. Here, $w$ represents an array of weights that is the same size as the array of features and provides a weight for each of the features when finally computing the label prediction in the form:

$$
prediction = w^\top \cdot\ x
$$

where $w$ is the final array of weights computed by the gradient descent, $x$ is the array of features of the observation point and $prediction$ is the label we predict should be associated to this observation.

The given `Helper` class implements the helper methods:
* `dot`: implements the dot product of two vectors and the dot product of a vector and a scalar
* `sum`: implements addition of two vectors
* `fill`: creates a vector of predefined size and initialize it with the predefined value

What you need to do is to implement the methods of the Linear Regresstion class `LR`, which are
* `rmsd`: computes the Root Mean Square Error of a given RDD of tuples of (label, prediction) using the formula:
$$
rmse = \sqrt{\frac{\sum\limits_{i=1}^n (label - prediction)^2}{n}}
$$
* `gradientSummand`: computes the following formula:
$$
gs_{ij} = (w_i^\top x_j - y_j)x_j
$$
* `gradient`: computes the following formula:
$$
gradient = \sum\limits_{j=1}^n gs_{ij}
$$

In [35]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

case class Instance(label: Double, features: Vector)

object Helper extends Serializable {
  def dot(v1: Vector, v2: Vector): Double = {
    val m = Matrices.dense(1, v1.size, v1.toArray)
    m.multiply(v2).values(0)
  }

  def dot(v: Vector, s: Double): Vector = {
    val baseArray = v.toArray.map(vi => vi * s)
    Vectors.dense(baseArray)
  }

  def sumVectors(v1: Vector, v2: Vector): Vector = {
    val baseArray = ((v1.toArray) zip (v2.toArray)).map { case (val1, val2) => val1 + val2 }
    Vectors.dense(baseArray)
  }

  def fillVector(size: Int, fillVal: Double): Vector = Vectors.dense(Array.fill[Double](size)(fillVal));
}

defined class Instance
defined object Helper


In [36]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

class LR() extends Serializable {
  def calcRMSE(labelsAndPreds: RDD[(Double, Double)]): Double = {
    val regressionMetrics = new RegressionMetrics(labelsAndPreds)
    regressionMetrics.rootMeanSquaredError
  }
  
  def gradientSummand(weights: Vector, lp: Instance): Vector = {
    val mult = (Helper.dot(weights, lp.features) - lp.label)
    val seq = (0 to lp.features.size - 1).map(i => lp.features(i) * mult)
    return Vectors.dense(seq.toArray)
  }
  
  def linregGradientDescent(trainData: RDD[Instance], numIters: Int): (Vector, Array[Double]) = {
    val n = trainData.count()
    val d = trainData.take(1)(0).features.size
    var w = Helper.fillVector(d, 0)
    val alpha = 1.0
    val errorTrain = Array.fill[Double](numIters)(0.0)

    for (i <- 0 until numIters) {
      val labelsAndPredsTrain = trainData.map(lp => (lp.label, Helper.dot(w, lp.features)))
      errorTrain(i) = calcRMSE(labelsAndPredsTrain)

      val gradient = trainData.map(lp => gradientSummand(w, lp)).reduce((v1, v2) => Helper.sumVectors(v1, v2))
      val alpha_i = alpha / (n * scala.math.sqrt(i + 1))
      val wAux = Helper.dot(gradient, (-1) * alpha_i)
      w = Helper.sumVectors(w, wAux)
    }
    (w, errorTrain)
  }
}

defined class LR


In [37]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

abstract class MyLinearModel[FeaturesType, Model <: MyLinearModel[FeaturesType, Model]]
  extends PredictionModel[FeaturesType, Model] {
}

class MyLinearModelImpl(override val uid: String, val weights: Vector, val trainingError: Array[Double])
    extends MyLinearModel[Vector, MyLinearModelImpl] {

  override def copy(extra: ParamMap): MyLinearModelImpl = defaultCopy(extra)

  def predict(features: Vector): Double = {
    println("Predicting")
    val prediction = Helper.dot(weights, features)
    prediction
  }
}

defined class MyLinearModel
defined class MyLinearModelImpl


In [38]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._

import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

abstract class MyLinearRegression[
    FeaturesType,
    Learner <: MyLinearRegression[FeaturesType, Learner, Model],
    Model <: MyLinearModel[FeaturesType, Model]]
  extends Predictor[FeaturesType, Learner, Model] {
}

class MyLinearRegressionImpl(override val uid: String)
    extends MyLinearRegression[Vector, MyLinearRegressionImpl, MyLinearModelImpl] {
  def this() = this(Identifiable.randomUID("linReg"))

  override def copy(extra: ParamMap): MyLinearRegressionImpl = defaultCopy(extra)
  
  def train(dataset: Dataset[_]): MyLinearModelImpl = {
    println("Training")

    val numIters = 10

    val instances: RDD[Instance] = dataset.select(
      col($(labelCol)), col($(featuresCol))).rdd.map {
        case Row(label: Double, features: Vector) =>
          Instance(label, features)
      }

    val (weights, trainingError) = new LR().linregGradientDescent(instances, numIters)

    new MyLinearModelImpl(uid, weights, trainingError)
  }
}

defined class MyLinearRegression
defined class MyLinearRegressionImpl


In [39]:
import org.apache.spark.ml.evaluation.RegressionEvaluator

val lr = new MyLinearRegressionImpl().setLabelCol("label").setFeaturesCol("features")
val model = lr.fit(trainSet)
val predictions = model.transform(trainSet)
predictions.select("prediction", "label", "features").show(5)

// Select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

Training
+--------------------+--------+--------------------+
|          prediction|   label|            features|
+--------------------+--------+--------------------+
|-8.61666156544522...| 94600.0|[-62.065401082150...|
|-8.62816182756746...| 85800.0|[-62.040445150874...|
|-8.62944646732868...|103600.0|[-62.040445150874...|
|-8.61615998040114...| 79000.0|[-62.025471592109...|
|-8.61378245178018...| 76100.0|[-62.015489219599...|
+--------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 8.236973861524635E37


lr = linReg_fdc40304a73d
model = linReg_fdc40304a73d
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_c53590e7b775
rmse = 8.236973861524635E37


8.236973861524635E37

---
# 9. An End-to-End Classification Test
As the last step, you are given a dataset called `data/ccdefault.csv`. The dataset represents default of credit card clients. It has 30,000 cases and 24 different attributes. More details about the dataset is available at `data/ccdefault.txt`. In this task you should make three models, compare their results and conclude the ideal solution. Here are the suggested steps:
1. Load the data.
2. Carry out some exploratory analyses (e.g., how various features and the target variable are distributed).
3. Train a model to predict the target variable (risk of `default`).
  - Employ three different models (logistic regression, decision tree, and random forest).
  - Compare the models' performances (e.g., AUC).
  - Defend your choice of best model (e.g., what are the strength and weaknesses of each of these models?).
4. What more would you do with this data? Anything to help you devise a better solution?

In [75]:
val ccdefault = spark.read.format("csv").option("header","true").option("inferSchema","true").load("data/ccdefault.csv")

ccdefault.show(5)


+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|DEFAULT|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|      1|
|  2|   120000|  2|        2|       2| 26|   -1|    2|    0|    0|    0|    2|     2682|     1725|     2682|     3272|     3455|     3261|       0|    1000|    1000|    1000|       0|    2000|    

ccdefault = [ID: int, LIMIT_BAL: int ... 23 more fields]


lastException: Throwable = null


[ID: int, LIMIT_BAL: int ... 23 more fields]

In [76]:
import org.apache.spark.sql.types.{IntegerType, StringType}
// Printing the schema to check if it is well inferred 
println(ccdefault.schema)

println("Number of rows are: " + ccdefault.count)

println("The number of parameters including the index are: " + ccdefault.columns.size)

println("Number of records with a credit higher than 200000 are: " + ccdefault.where("LIMIT_BAL > 200000").count)

println("The percentage of females: " + (((ccdefault.where("SEX == 2").count.toDouble)/(ccdefault.count.toDouble))*100).toString.substring(0,4) + "%")

println("The percentage of males: " + (((ccdefault.where("SEX == 1").count.toDouble)/(ccdefault.count.toDouble))*100).toString.substring(0,4) + "%")



println("Average Credit By Gender")
val sexData = Seq(
  Row(1, "Male"),
  Row(2, "Female")
)

val sexSchema = List(
  StructField("number", IntegerType, true),
  StructField("SEX DISP", StringType, true)
)

val sexDF = spark.createDataFrame(
  spark.sparkContext.parallelize(sexData),
  StructType(sexSchema)
)
val sexDisp = ccdefault.groupBy("SEX").agg(avg("LIMIT_BAL").alias("AVG LIMIT BAL")).join(sexDF, $"SEX" === $"number")
sexDisp.select($"SEX DISP",$"AVG LIMIT BAL").orderBy(desc("AVG LIMIT BAL")).show()

println("Average Credit By Marital Status")
val marData = Seq(
  Row(1, "Married"),
  Row(2, "Single"),
  Row(3, "Other"),
  Row(0, "Unknown")
)

val marSchema = List(
  StructField("number", IntegerType, true),
  StructField("MARRIAGE DISP", StringType, true)
)

val marDF = spark.createDataFrame(
  spark.sparkContext.parallelize(marData),
  StructType(marSchema)
)
val marDisp = ccdefault.groupBy("MARRIAGE").agg(avg("LIMIT_BAL").alias("AVG LIMIT BAL")).join(marDF, $"MARRIAGE" === $"number")
marDisp.select($"MARRIAGE DISP",$"AVG LIMIT BAL").orderBy(desc("AVG LIMIT BAL")).show()

println("Average Credit By Education")
val edData = Seq(
  Row(1, "Graduate School"),
    Row(2, "University"),
    Row(3, "High School"),
    Row(4, "Others"),
  Row(0, "Unknown0"),
    Row(5, "Unknown5"),
    Row(6, "Unknow6")
)

val edSchema = List(
  StructField("number", IntegerType, true),
  StructField("EDUCATION DISP", StringType, true)
)

val edDF = spark.createDataFrame(
  spark.sparkContext.parallelize(edData),
  StructType(edSchema)
)
val edDisp = ccdefault.groupBy("EDUCATION").agg(avg("LIMIT_BAL").alias("AVG LIMIT BAL")).join(edDF, $"EDUCATION" === $"number")
edDisp.select($"EDUCATION DISP",$"AVG LIMIT BAL").orderBy(desc("AVG LIMIT BAL")).show()

val avgAge = ccdefault.agg(avg("AGE")).collect.mkString
val maxAge = ccdefault.agg(max("AGE")).collect.mkString
val minAge = ccdefault.agg(min("AGE")).collect.mkString

println("The average age is: " + avgAge.substring(1,avgAge.size-1))
println("The max age is: " + maxAge.substring(1,maxAge.size-1))
println("The min age is: " + minAge.substring(1,minAge.size-1))


StructType(StructField(ID,IntegerType,true), StructField(LIMIT_BAL,IntegerType,true), StructField(SEX,IntegerType,true), StructField(EDUCATION,IntegerType,true), StructField(MARRIAGE,IntegerType,true), StructField(AGE,IntegerType,true), StructField(PAY_0,IntegerType,true), StructField(PAY_2,IntegerType,true), StructField(PAY_3,IntegerType,true), StructField(PAY_4,IntegerType,true), StructField(PAY_5,IntegerType,true), StructField(PAY_6,IntegerType,true), StructField(BILL_AMT1,IntegerType,true), StructField(BILL_AMT2,IntegerType,true), StructField(BILL_AMT3,IntegerType,true), StructField(BILL_AMT4,IntegerType,true), StructField(BILL_AMT5,IntegerType,true), StructField(BILL_AMT6,IntegerType,true), StructField(PAY_AMT1,IntegerType,true), StructField(PAY_AMT2,IntegerType,true), StructField(PAY_AMT3,IntegerType,true), StructField(PAY_AMT4,IntegerType,true), StructField(PAY_AMT5,IntegerType,true), StructField(PAY_AMT6,IntegerType,true), StructField(DEFAULT,IntegerType,true))
Number of rows a

sexData = List([1,Male], [2,Female])
sexSchema = List(StructField(number,IntegerType,true), StructField(SEX DISP,StringType,true))
sexDF = [number: int, SEX DISP: string]
sexDisp = [SEX: int, AVG LIMIT BAL: double ... 2 more fields]
marData = List([1,Married], [2,Single], [3,Other], [0,Unknown])
marSchema = List(StructField(number,IntegerType,true), StructField(MARRIAGE DISP,StringType,true))
marDF = [number: int, MARRIAGE DISP: string]
marDisp = [MAR...


[MAR...

In [77]:
val colNum = ccdefault.columns


// THERE ARE NO MISSING VALUES
for (c <- colNum) {
    println("Column " + c + " has " + ccdefault.filter(c + " is null").count + " missing values.")
}

Column ID has 0 missing values.
Column LIMIT_BAL has 0 missing values.
Column SEX has 0 missing values.
Column EDUCATION has 0 missing values.
Column MARRIAGE has 0 missing values.
Column AGE has 0 missing values.
Column PAY_0 has 0 missing values.
Column PAY_2 has 0 missing values.
Column PAY_3 has 0 missing values.
Column PAY_4 has 0 missing values.
Column PAY_5 has 0 missing values.
Column PAY_6 has 0 missing values.
Column BILL_AMT1 has 0 missing values.
Column BILL_AMT2 has 0 missing values.
Column BILL_AMT3 has 0 missing values.
Column BILL_AMT4 has 0 missing values.
Column BILL_AMT5 has 0 missing values.
Column BILL_AMT6 has 0 missing values.
Column PAY_AMT1 has 0 missing values.
Column PAY_AMT2 has 0 missing values.
Column PAY_AMT3 has 0 missing values.
Column PAY_AMT4 has 0 missing values.
Column PAY_AMT5 has 0 missing values.
Column PAY_AMT6 has 0 missing values.
Column DEFAULT has 0 missing values.


colNum = Array(ID, LIMIT_BAL, SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6, DEFAULT)


Array(ID, LIMIT_BAL, SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6, DEFAULT)

In [ ]:
// Create the correlational matrix
var matrix = Array.ofDim[Double](colNum.size,colNum.size)
var i = 0

for (c <- colNum) {
    var j = 0
    for (ci <- colNum) 
    {
        matrix(i)(j) = ccdefault.stat.corr(c, ci, "pearson").toDouble
        j += 1
    }
    i += 1
}
// Values imported to matlab to print out the heatmap displayed in the rport

In [81]:
// Rescaling for faster gradient descent 


import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}


val colfiltered = ccdefault.columns.filter(_ != "ID").filter(_ != "MARRIAGE").filter(_ != "AGE").filter(_ != "SEX")
.filter(_ != "EDUCATION").filter(_ != "DEFAULT")
.filter(_ != "PAY_0").filter(_ != "PAY_2").filter(_ != "PAY_3").filter(_ != "PAY_4")
.filter(_ != "PAY_5").filter(_ != "PAY_6")



val va = new VectorAssembler().setInputCols(colfiltered).setOutputCol("features1")
val vectoredCC = va.transform(ccdefault)

val scaler = new StandardScaler().setInputCol("features1").setOutputCol("scales")
val scaledDefault = scaler.fit(vectoredCC).transform(vectoredCC)


val va2 = new VectorAssembler().setInputCols(Array("scales", "MARRIAGE", "AGE", "SEX", "EDUCATION",
                                                  "PAY_0", "PAY_2", "PAY_3", "PAY_4", "PAY_5"
                                                  ,"PAY_6")).setOutputCol("features")

val dataset = va2.transform(scaledDefault).select("features", "DEFAULT")

dataset.show(5)



+--------------------+-------+
|            features|DEFAULT|
+--------------------+-------+
|[0.15414535998894...|      1|
|[0.92487215993365...|      1|
|[0.69365411995024...|      0|
|[0.38536339997235...|      0|
|[0.38536339997235...|      0|
+--------------------+-------+
only showing top 5 rows



colfiltered = Array(LIMIT_BAL, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6)
va = vecAssembler_e6cbe528bd90
vectoredCC = [ID: int, LIMIT_BAL: int ... 24 more fields]
scaler = stdScal_6b0bfa49b99c
scaledDefault = [ID: int, LIMIT_BAL: int ... 25 more fields]
va2 = vecAssembler_779138040be0
dataset = [features: vector, DEFAULT: int]


lastException: Throwable = null


[features: vector, DEFAULT: int]

In [99]:
// Divide training and test set 

var Array(trainSet1, testSet1) = dataset.randomSplit(Array(0.7, 0.3))

testSet1.show(5)


+--------------------+-------+
|            features|DEFAULT|
+--------------------+-------+
|(23,[0,1,2,3,4,5,...|      0|
|(23,[0,1,2,3,4,5,...|      0|
|(23,[0,1,2,3,4,5,...|      0|
|(23,[0,1,2,3,4,5,...|      0|
|(23,[0,1,2,3,4,5,...|      0|
+--------------------+-------+
only showing top 5 rows



trainSet1 = [features: vector, DEFAULT: int]
testSet1 = [features: vector, DEFAULT: int]


[features: vector, DEFAULT: int]

In [101]:
// Logistic Regression

import org.apache.spark.ml.classification.LogisticRegression

import org.apache.spark.ml.classification.BinaryLogisticRegressionSummary
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator




var lr = new LogisticRegression()
  .setMaxIter(100)
  .setRegParam(0)
  .setElasticNetParam(0.8)
  .setFeaturesCol("features")
  .setLabelCol("DEFAULT")

testSet1.show(5)


var lrModel = lr.fit(trainSet1)

var predictions = lrModel.transform(testSet1)

predictions.select("prediction", "DEFAULT", "features").show(100)


// select (prediction, true label) and compute test error
val evaluator = new MulticlassClassificationEvaluator()
  .setLabelCol("DEFAULT")
  .setPredictionCol("prediction")
  .setMetricName("accuracy")
val accuracy = evaluator.evaluate(predictions)
//println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

//val accuracy = evaluator.evaluate(predictions)
println("Test Error = " + (1.0 - accuracy))


val trainingSummary = lrModel.summary


val binarySummary = trainingSummary.asInstanceOf[BinaryLogisticRegressionSummary]


val roc = binarySummary.roc
roc.show()
println(s"areaUnderROC: ${binarySummary.areaUnderROC}")


+--------------------+-------+
|            features|DEFAULT|
+--------------------+-------+
|(23,[0,1,2,3,4,5,...|      0|
|(23,[0,1,2,3,4,5,...|      0|
|(23,[0,1,2,3,4,5,...|      0|
|(23,[0,1,2,3,4,5,...|      0|
|(23,[0,1,2,3,4,5,...|      0|
+--------------------+-------+
only showing top 5 rows

+----------+-------+--------------------+
|prediction|DEFAULT|            features|
+----------+-------+--------------------+
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       1.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,

lr = logreg_448b3fca7c39
lrModel = logreg_448b3fca7c39
predictions = [features: vector, DEFAULT: int ... 3 more fields]
evaluator = mcEval_21ef736670d7
accuracy = 0.8089143370382845
trainingSummary = org.apache.spark.ml.classification.BinaryLogisticRegressionTrainingSummaryImpl@48bca9dd


lastException: Throwable = null
binarySumma...


org.apache.spark.ml.classification.BinaryLogisticRegressionTrainingSummaryImpl@48bca9dd

In [143]:
// Hyperparameter Optimization (number of iterations, regularization and treshold)

import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.RegressionEvaluator
import org.apache.spark.ml.tuning.CrossValidator

import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator



val paramGrid = new ParamGridBuilder().addGrid(lr.threshold, Array(0.4, 0.5, 0.6)).addGrid(lr.maxIter, Array(10, 100, 1000))
.addGrid(lr.regParam, Array(0, 0.1, 0.2, 0.3)).build()

val evaluator = new MulticlassClassificationEvaluator()
  .setLabelCol("DEFAULT")
  .setPredictionCol("prediction")
  .setMetricName("accuracy")
val cv = new CrossValidator().setEstimator(lr).setEvaluator(evaluator).setEstimatorParamMaps(paramGrid).setNumFolds(3)
val cvModel = cv.fit(trainSet1)

val predictions = cvModel.transform(testSet1)
predictions.select("prediction", "DEFAULT", "features").show(5)

// Printing the chosen hyperparameters:

val bestmodel = cvModel.bestModel

println("Best Parameters:")

println(bestmodel.parent.extractParamMap())


val accuracy = evaluator.evaluate(predictions)
println("Test Error = " + (1.0 - accuracy))



Name: Unknown Error
Message: <console>:279: error: value summary is not a member of org.apache.spark.ml.Model[_$7]
       val trainingSummary = cvModel.bestModel.summary
                                               ^

StackTrace: 

In [139]:
// Decision Trees

import org.apache.spark.ml.classification.DecisionTreeClassifier
import org.apache.spark.ml.classification.DecisionTreeClassificationModel
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator
import org.apache.spark.ml.classification.BinaryLogisticRegressionSummary
import org.apache.spark.mllib.evaluation.BinaryClassificationMetrics


val dt = new DecisionTreeClassifier().setLabelCol("DEFAULT").setFeaturesCol("features")

// train the model
val dtModel = dt.fit(trainSet1)

// make predictions on the test data
val predictions = dtModel.transform(testSet1)
//predictions.select("prediction", "DEFAULT", "features").show(100)

// select (prediction, true label) and compute test error
val evaluator = new MulticlassClassificationEvaluator()
  .setLabelCol("DEFAULT")
  .setPredictionCol("prediction")
  .setMetricName("accuracy")
val accuracy = evaluator.evaluate(predictions)


val convert = predictions.rdd

val converted = convert.map { 
  row: Row => (row.get(1).toString.toDouble, row.get(4).toString.toDouble)
}


val metrics = new BinaryClassificationMetrics(converted)

val roc = metrics.roc

// AUROC
val auROC = metrics.areaUnderROC
println(s"Area under ROC = $auROC")


println("Test Error = " + (1.0 - accuracy))



+--------------------+-------+--------------+--------------------+----------+
|            features|DEFAULT| rawPrediction|         probability|prediction|
+--------------------+-------+--------------+--------------------+----------+
|(23,[0,1,2,3,4,5,...|      0|[3788.0,546.0]|[0.87401938163359...|       0.0|
|(23,[0,1,2,3,4,5,...|      0|[3788.0,546.0]|[0.87401938163359...|       0.0|
|(23,[0,1,2,3,4,5,...|      0|[1313.0,399.0]|[0.76693925233644...|       0.0|
|(23,[0,1,2,3,4,5,...|      0|[5605.0,492.0]|[0.91930457602099...|       0.0|
|(23,[0,1,2,3,4,5,...|      0|[5605.0,492.0]|[0.91930457602099...|       0.0|
+--------------------+-------+--------------+--------------------+----------+
only showing top 5 rows

Area under ROC = 0.7603970098796812
Test Error = 0.18041989446502749


dt = dtc_ef26b5155a42
dtModel = DecisionTreeClassificationModel (uid=dtc_ef26b5155a42) of depth 5 with 61 nodes
predictions = [features: vector, DEFAULT: int ... 3 more fields]
evaluator = mcEval_f473788ce7ee


accuracy:...


mcEval_f473788ce7ee

In [145]:
// Random Forest

import org.apache.spark.ml.classification.{RandomForestClassificationModel, RandomForestClassifier}
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator

val rf = new RandomForestClassifier().setLabelCol("DEFAULT").setFeaturesCol("features")

// train the model
val rfModel = rf.fit(trainSet1)

// make predictions on the test data
val predictions = rfModel.transform(testSet1)
predictions.select("prediction", "DEFAULT", "features").show(1000)

// select (prediction, true label) and compute test error
val evaluator = new MulticlassClassificationEvaluator()
  .setLabelCol("DEFAULT")
  .setPredictionCol("prediction")
  .setMetricName("accuracy")
val accuracy = evaluator.evaluate(predictions)
//println(s"Root Mean Squared Error (RMSE) on test data = $rmse")


val convert = predictions.rdd

val converted = convert.map { 
  row: Row => (row.get(1).toString.toDouble, row.get(4).toString.toDouble)
}


val metrics = new BinaryClassificationMetrics(converted)

val roc = metrics.roc

// AUROC
val auROC = metrics.areaUnderROC
println(s"Area under ROC = $auROC")


println("Test Error = " + (1.0 - accuracy))

//val accuracy = evaluator.evaluate(predictions)
println("Test Error = " + (1.0 - accuracy))


+----------+-------+--------------------+
|prediction|DEFAULT|            features|
+----------+-------+--------------------+
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       1.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4

Area under ROC = 0.7534621176568613
Test Error = 0.1826653194116986
Test Error = 0.1826653194116986


rf = rfc_47de8725aa40
rfModel = RandomForestClassificationModel (uid=rfc_47de8725aa40) with 20 trees
predictions = [features: vector, DEFAULT: int ... 3 more fields]
evaluator = mcEval_42e1616e0988
accuracy = 0.8173346805883014
convert = MapPartitionsRDD[19845] at rdd at <console>:264


lastException: Throwable = null
converted: org.apache.spark.rdd.RDD[(Double, Double)] ...


MapPartitionsRDD[19845] at rdd at <console>:264

In [59]:
// Hyperparamater tuning



import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.RegressionEvaluator
import org.apache.spark.ml.tuning.CrossValidator

val paramGrid = new ParamGridBuilder().addGrid(rf.numTrees, Array(1, 5, 10)).addGrid(rf.maxDepth, Array(5, 10, 15)).build()

val evaluator = new MulticlassClassificationEvaluator()
  .setLabelCol("DEFAULT")
  .setPredictionCol("prediction")
  .setMetricName("accuracy")
val cv = new CrossValidator().setEstimator(rf).setEvaluator(evaluator).setEstimatorParamMaps(paramGrid).setNumFolds(3)
val cvModel = cv.fit(trainSet)

val predictions = cvModel.transform(testSet)
predictions.select("prediction", "DEFAULT", "features").show(5)


// Printing the chosen hyperparameters:

val bestmodel = cvModel.bestModel

println("Best Parameters:")

println(bestmodel.parent.extractParamMap())


val accuracy = evaluator.evaluate(predictions)
println("Test Error = " + (1.0 - accuracy))

+----------+-------+--------------------+
|prediction|DEFAULT|            features|
+----------+-------+--------------------+
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
|       0.0|      1|(23,[0,1,2,3,4,5,...|
|       0.0|      0|(23,[0,1,2,3,4,5,...|
+----------+-------+--------------------+
only showing top 5 rows

Best Parameters:
{
	rfc_f6e0b82f98b0-cacheNodeIds: false,
	rfc_f6e0b82f98b0-checkpointInterval: 10,
	rfc_f6e0b82f98b0-featureSubsetStrategy: auto,
	rfc_f6e0b82f98b0-featuresCol: features,
	rfc_f6e0b82f98b0-impurity: gini,
	rfc_f6e0b82f98b0-labelCol: DEFAULT,
	rfc_f6e0b82f98b0-maxBins: 32,
	rfc_f6e0b82f98b0-maxDepth: 5,
	rfc_f6e0b82f98b0-maxMemoryInMB: 256,
	rfc_f6e0b82f98b0-minInfoGain: 0.0,
	rfc_f6e0b82f98b0-minInstancesPerNode: 1,
	rfc_f6e0b82f98b0-numTrees: 1,
	rfc_f6e0b82f98b0-predictionCol: prediction,
	rfc_f6e0b82f98b0-probabilityCol: probability,
	rfc_f6e0b82f98b0-rawPredictionCol: r

paramGrid = 


Array({
	rfc_f6e0b82f98b0-maxDepth: 5,
	rfc_f6e0b82f98b0-numTrees: 1
}, {
	rfc_f6e0b82f98b0-maxDepth: 5,
	rfc_f6e0b82f98b0-numTrees: 5
}, {
	rfc_f6e0b82f98b0-maxDepth: 5,
	rfc_f6e0b82f98b0-numTrees: 10
}, {
	rfc_f6e0b82f98b0-maxDepth: 10,
	rfc_f6e0b82f98b0-numTrees: 1
}, {
	rfc_f6e0b82f98b0-maxDepth: 10,
	rfc_f6e0b82f98b0-numTrees: 5
}, {
	rfc_f6e0b82f98b0-maxDepth: 10,
	rfc_f6e0b82f98b0-numTrees: 10
}, {
	rfc_f6e0b82f98b0-maxDepth: 15,
	rfc_f6e0b82f98b0-numTrees: 1
}, {
	rfc_f6e0b82f98b0-maxDepth: 15,
	rfc_f6e0b82f98b0-numTrees: 5
}, {
	rfc_f6e0b82f98b0-maxDepth: 15,
	rfc_f6e0...


In [60]:
// Result 



Name: Syntax Error.
Message: 
StackTrace: 

In [61]:
val ccdefault1 = ccdefault.withColumn("PAYBILL1", ccdefault("PAY_AMT1") / ccdefault("BILL_AMT1"))
val ccdefault2 = ccdefault1.withColumn("PAYBILL2", ccdefault1("PAY_AMT2") / ccdefault1("BILL_AMT2"))
val ccdefault3 = ccdefault2.withColumn("PAYBILL3", ccdefault2("PAY_AMT3") / ccdefault2("BILL_AMT3"))
val ccdefault4 = ccdefault3.withColumn("PAYBILL4", ccdefault3("PAY_AMT4") / ccdefault3("BILL_AMT4"))
val ccdefault5 = ccdefault4.withColumn("PAYBILL5", ccdefault4("PAY_AMT5") / ccdefault4("BILL_AMT5"))

// 
val ccdefaultFinal = ccdefault5.na.fill(0, Seq("PAYBILL1", "PAYBILL2", "PAYBILL3", "PAYBILL4", "PAYBILL5"))
ccdefaultFinal.select("PAYBILL1", "PAYBILL2", "PAYBILL3", "PAYBILL4", "PAYBILL5").show(5)

+-------------------+-------------------+--------------------+-------------------+-------------------+
|           PAYBILL1|           PAYBILL2|            PAYBILL3|           PAYBILL4|           PAYBILL5|
+-------------------+-------------------+--------------------+-------------------+-------------------+
|                0.0|0.22211476466795615|                 0.0|                0.0|                0.0|
|                0.0| 0.5797101449275363| 0.37285607755406414| 0.3056234718826406|                0.0|
|0.05191696022435788|0.10693662222855921|  0.0737517516041006|0.06977880120019538| 0.0668985817500669|
|0.04256224728665674|0.04185930794269484|0.024345215150838897|0.03885003885003885|0.03691425808902241|
|0.23209933851688522|  6.469312169312169| 0.27905678805636946| 0.4297994269340974|0.03598662906090045|
+-------------------+-------------------+--------------------+-------------------+-------------------+
only showing top 5 rows



ccdefault1 = [ID: int, LIMIT_BAL: int ... 24 more fields]
ccdefault2 = [ID: int, LIMIT_BAL: int ... 25 more fields]
ccdefault3 = [ID: int, LIMIT_BAL: int ... 26 more fields]
ccdefault4 = [ID: int, LIMIT_BAL: int ... 27 more fields]
ccdefault5 = [ID: int, LIMIT_BAL: int ... 28 more fields]
ccdefaultFinal = [ID: int, LIMIT_BAL: int ... 28 more fields]


[ID: int, LIMIT_BAL: int ... 28 more fields]

In [62]:

ccdefault5.na.fill(0,Seq("PAYBILL5")).show


+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+-------------------+--------------------+--------------------+--------------------+--------------------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|DEFAULT|           PAYBILL1|            PAYBILL2|            PAYBILL3|            PAYBILL4|            PAYBILL5|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+-------------------+--------------------+--------------------+--------------------+--------------------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     39

In [66]:



///PCA Analysis 


import org.apache.spark.ml.feature.PCA
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

var colfilteredPCA = ccdefault.columns.filter(_ != "DEFAULT").filter(_ != "ID")

var vaPCA = new VectorAssembler().setInputCols(colfilteredPCA).setOutputCol("featuresPrePCA")
var datasetPCA = vaPCA.transform(ccdefault).select("featuresPrePCA")

var pcam = new PCA().setK(colfilteredPCA.size).setInputCol("featuresPrePCA").setOutputCol("featuresAfterPCA")
var model = pcam.fit(datasetPCA)

println("Explained variance:")
println(model.explainedVariance)

import org.apache.spark.ml.feature.PCA
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

colfilteredPCA = ccdefault.columns.filter(_ != "DEFAULT").filter(_ != "ID")

vaPCA = new VectorAssembler().setInputCols(colfilteredPCA).setOutputCol("featuresPrePCA")
datasetPCA = vaPCA.transform(ccdefault).select("featuresPrePCA")

pcam = new PCA().setK(6).setInputCol("featuresPrePCA").setOutputCol("featuresAfterPCA")
var modelcam = pcam.fit(datasetPCA)

var result = modelcam.transform(datasetPCA).select("featuresAfterPCA");
result.show(false);

Explained variance:
[0.6104370130775205,0.2953538127654238,0.030524192486377534,0.01692859259388121,0.009420419259578367,0.00904174544504892,0.007544457903494719,0.006384813792058376,0.005837087707793788,0.002966714115486337,0.0023850600245916296,0.001697743534793529,0.0014783452973830978,1.8316233690862943E-9,9.377458096985944E-11,1.9425662598313225E-11,1.2476156675058675E-11,1.1781891888215633E-11,7.83531976220339E-12,5.7763076397999845E-12,5.372751023083941E-12,4.6441490647924875E-12,3.857813453068243E-12]
+----------------------------------------------------------------------------------------------------------------------+
|featuresAfterPCA                                                                                                      |
+----------------------------------------------------------------------------------------------------------------------+
|[-12812.786808217688,-15670.275813184824,-3695.229912793964,442.3103274096919,454.71556734703955,-377.21426893473665] |
|

colfilteredPCA = Array(LIMIT_BAL, SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6)
vaPCA = vecAssembler_d1f005d4b405
datasetPCA = [featuresPrePCA: vector]
pcam = pca_44ce076e1982
model = pca_c959e7888c01


pca_c959e7888c01

In [147]:
/// RANDOM FOREST AFTER PCA AND DIMENSIONALITY REDUCTION 

// Combining the reduced number of features and their values with the labels

import org.apache.spark.sql.functions.monotonically_increasing_id
val resDes = result.withColumn("id", monotonically_increasing_id())
    .join(ccdefault.withColumn("id", monotonically_increasing_id()), Seq("id"))
    .drop("id")
    
resDes.select("featuresAfterPCA","DEFAULT")show



val Array(trainSet, testSet) = resDes.randomSplit(Array(0.7, 0.3))





import org.apache.spark.ml.classification.{RandomForestClassificationModel, RandomForestClassifier}
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator

val rf = new RandomForestClassifier().setLabelCol("DEFAULT").setFeaturesCol("featuresAfterPCA")

// train the model
val rfModel = rf.fit(trainSet)

// make predictions on the test data
val predictions = rfModel.transform(testSet)
predictions.select("prediction", "DEFAULT", "featuresAfterPCA").show(1000)

// select (prediction, true label) and compute test error
val evaluator = new MulticlassClassificationEvaluator()
  .setLabelCol("DEFAULT")
  .setPredictionCol("prediction")
  .setMetricName("accuracy")
val accuracy = evaluator.evaluate(predictions)

//https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.mllib.evaluation.BinaryClassificationMetrics
//https://spark.apache.org/docs/latest/mllib-evaluation-metrics.html
val convert = predictions.rdd

val converted = convert.map { 
  row: Row => (row.get(1).toString.toDouble, row.get(4).toString.toDouble)
}


val metrics = new BinaryClassificationMetrics(converted)

val roc = metrics.roc

// AUROC
val auROC = metrics.areaUnderROC
println(s"Area under ROC = $auROC")
println("Test Error = " + (1.0 - accuracy))



+--------------------+-------+
|    featuresAfterPCA|DEFAULT|
+--------------------+-------+
|[-12812.786808217...|      1|
|[-65074.001077343...|      1|
|[-80868.615713359...|      0|
|[-108070.30199051...|      0|
|[-64466.359528672...|      0|
|[-112237.46519070...|      0|
|[-1206503.3221489...|      0|
|[-54458.602697405...|      0|
|[-92443.201913202...|      0|
|[-18638.215498517...|      0|
|[-111178.18959032...|      0|
|[-160620.66538914...|      0|
|[-325163.00812947...|      0|
|[-156568.92195420...|      1|
|[-256299.74092488...|      0|
|[-94814.191463392...|      0|
|[-47350.304727247...|      1|
|[-517339.99178077...|      0|
|[-176972.63749279...|      0|
|[-88486.318767707...|      0|
+--------------------+-------+
only showing top 20 rows

+----------+-------+--------------------+
|prediction|DEFAULT|    featuresAfterPCA|
+----------+-------+--------------------+
|       0.0|      0|[-1484768.0996714...|
|       0.0|      0|[-1479391.5612273...|
|       0.0|      1|

Area under ROC = 0.4584381097990304
Test Error = 0.21693004751906286
Test Error = 0.21693004751906286


resDes = [featuresAfterPCA: vector, LIMIT_BAL: int ... 23 more fields]
trainSet = [featuresAfterPCA: vector, LIMIT_BAL: int ... 23 more fields]
testSet = [featuresAfterPCA: vector, LIMIT_BAL: int ... 23 more fields]
rf = rfc_f95cbcdd1d24
rfModel = RandomForestClassificationModel (uid=rfc...


RandomForestClassificationModel (uid=rfc...